rust/lance/src/dataset.rs

```diff
@@ -915,14 +915,20 @@ impl Dataset {
     ///
     /// It offers a fast path of counting rows by just computing via metadata.
     #[instrument(skip_all)]
-    pub async fn count_rows(&self) -> Result<usize> {
-        // Open file to read metadata.
-        let counts = stream::iter(self.get_fragments())
-            .map(|f| async move { f.count_rows().await })
-            .buffer_unordered(16)
-            .try_collect::<Vec<_>>()
-            .await?;
-        Ok(counts.iter().sum())
+    pub async fn count_rows(&self, filter: Option<String>) -> Result<usize> {
+        // TODO: consolidate the count_rows into Scanner plan.
+        if let Some(filter) = filter {
+            let mut scanner = self.scan();
+            scanner.filter(&filter)?;
```

Comment:
Could you double check we aren't loading more columns than necessary? (`dbg!()` the plan within `count_rows()`?) You might need to add:

```suggestion
            scanner
                .filter(&filter)?
                .with_row_id()
                .project::<&str>(&[])?;
```

Reply:
Got it.